cmake/build_env_info.cmake

```diff
@@ -0,0 +1,45 @@
+find_package(Git QUIET)
+
+if(Git_FOUND)
+    # Get the current working branch
+    execute_process(
+        COMMAND ${GIT_EXECUTABLE} rev-parse --abbrev-ref HEAD
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE GIT_BRANCH
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    # Get the latest commit hash
+    execute_process(
+        COMMAND ${GIT_EXECUTABLE} rev-parse --verify HEAD
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE GIT_COMMIT_HASH
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    # Get the latest commit description
+    execute_process(
+        COMMAND ${GIT_EXECUTABLE} show -s --format=%s
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE GIT_COMMIT_DESCRIPTION
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    # Get the latest commit date
+    execute_process(
+        COMMAND ${GIT_EXECUTABLE} log -1 --format=%cI
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE GIT_COMMIT_DATE
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    # Check if repo is dirty / clean
+    execute_process(
+        COMMAND ${GIT_EXECUTABLE} diff-index --quiet HEAD --
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        RESULT_VARIABLE IS_DIRTY
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
+    if(IS_DIRTY EQUAL 0)
+        set(GIT_CLEAN_STATUS "CLEAN")
+    else()
+        set(GIT_CLEAN_STATUS "DIRTY")
+    endif()
+
+    configure_file(cmake/Environment_Info.h.in Environment_Info.hpp @ONLY)
```

Comment:
The commit hash is something we will want to include in what that gets displayed with `Kokkos::print_configuration` (I am NOT suggesting to do that now).  Can you please discuss if you considered the behavior when re-building Kokkos.  If I edit the code and re-build does the dirty state or new commit get picked up?  If so, what will be rebuilt?  Anything that sees that `Environment_Info.hpp` header?  plus obviously link again?

Reply:
We worked through details of how to do this well in vt. The commit hash should go in a generated cpp file, that's only overwritten if it's changed. That way the rebuild is isolated, and relink only happens if there's a reason to